keps/sig-network/20201010-kube-proxy-v2/README.md

```diff
@@ -0,0 +1,753 @@
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+At the beginning, `kube-proxy` was designed to handle the translation of Service objects to OS-level resources.
+Implementations have been userland, then iptables, and now ipvs. With the growth of the Kubernetes project, more
+implementations came to life, for instance with eBPF, and often in relation to other goals (Calico to manage
+the network overlay, Cilium to manage app-level security, metallb to provide an external LB for bare-metal clusters, etc).
+
+Along this cambrian explosion of third-party software, the Service object itself received new concepts to improve the
+abstraction, for instance to express topology. This, third-party implementation are expected to update and become more
+complex over time, even if their core doesn't change (ie, the eBPF translation layer is not affected).
+
+This KEP is born from the conviction that more decoupling of the Service object and the actual implementations is required,
+by introducing an intermediate, node-level abstaction provider. This abstraction is expected to be the result of applying
+Kubernetes' `Service` semantics and business logic to a simpler, more stable API.
+
```

Comment:
Random additional point that I don't know exactly where it belongs in the doc: having this daemon would mean that proxy implementations wouldn't need to be ported to EndpointSlice, since the Endpoints vs EndpointSlices distinction would be handled entirely within the daemon...

Reply:
yes, good example of the interest of the decoupling I'm trying to achieve. Same idea with adding the topology concept.